torch/_lobpcg.py

```diff
@@ -0,0 +1,702 @@
+        """Return a transformation matrix that is used in Rayleigh-Ritz
+        procedure for reducing a general eigenvalue problem :math:`(S^TAS)
+        C = (S^TBS) C E` to a standard eigenvalue problem :math: `(Ri^T
+        S^TAS Ri) Z = Z E` where `C = Ri Z`.
+
+        .. note:: In the original Rayleight-Ritz procedure in
+          [DuerschEtal2018], the problem is formulated as follows::
+
+            SAS = S^T A S
+            SBS = S^T B S
+            D = (<diagonal matrix of SBS>) ** -1/2
+            R^T R = Cholesky(D SBS D)
+            Ri = D R^-1
+            solve symeig problem Ri^T SAS Ri Z = Theta Z
+            C = Ri Z
+
+          To reduce the number of matrix products (denoted by empty
+          space between matrices), here we introduce element-wise
+          products (denoted by symbol `*`) so that the Rayleight-Ritz
+          procedure becomes::
+
+            SAS = S^T A S
+            SBS = S^T B S
+            d = (<diagonal of SBS>) ** -1/2    # this is 1-d column vector
+            dd = d d^T                         # this is 2-d matrix
+            R^T R = Cholesky(dd * SBS)
+            Ri = R^-1 * d                      # broadcasting
+            solve symeig problem Ri^T SAS Ri Z = Theta Z
+            C = Ri Z
+
+          where `dd` is 2-d matrix that replaces matrix products `D M
+          D` with one element-wise product `M * dd`; and `d` replaces
+          matrix product `D M` with element-wise product `M *
+          d`. Also, creating the diagonal matrix `D` is avoided.
+
+        Arguments:
+        S (Tensor): the matrix basis for the search subspace, size is
+                    :math:`(m, n)`.
+
+        Returns:
+        Ri (tensor): upper-triangular transformation matrix of size
+                     :math:`(n, n)`.
+
+        """
+        B = self.B
+        mm = torch.matmul
+        SBS = _utils.qform(B, S)
+
+        d1 = SBS.diagonal(0, -2, -1) ** -0.5
+        d = torch.zeros((d1.shape[0], 1), dtype=d1.dtype, device=d1.device)
+        d[:, 0] = d1
+
+        dd = mm(d, _utils.transpose(d))
+        R = torch.cholesky(dd * SBS, upper=True)
+        # TODO: could use LAPACK ?trtri as R is upper-triangular
+        Rinv = torch.inverse(R)
+        return Rinv * d
+
+    def _get_svqb(self,
+                  U,     # Tensor
+                  drop,  # bool
+                  tau    # float
+                  ):
+        """Return B-orthonormal U.
+
+        .. note:: When `drop` is `False` then `svqb` is based on the
+                  Algorithm 4 from [DuerschPhD2015] that is a slight
+                  modification of the corresponding algorithm introduced
+                  in [StathopolousWu2002]. Otherwise,
+
+        Arguments:
+
+          U (Tensor) : initial approximation, size is (m, n)
+          drop (bool) : when True, drop columns that
+                     contribution to the `span([U])` is small.
+          tau (float) : positive tolerance
```

Comment:
Not clear what tolerance means in this context, unless you read the paper. Also, normally tolerance understood as an additive value, but in the paper it is a multiplicative factor.

Reply:
'contribution of a U column `U[i]` to the span(U) is small' means that the particular U column is nearly collinear to `span(<U without U[i]>)`.

Reply:
In L2-norm of the orthogonal projection residual?

Reply:
I think neither. See the source or the method paper for the details of the dropping criteria.

Reply:
I am closing this question as it is more of a theoretical kind and there is no real issue with the implementation. Feel free to reopen it when having more questions though.

Reply:
Yeah, it is related to the B-induced norm. The method is called conjugate for a good reason ;)